---
title: "Class 11: Structural Bioinformatics pt2"
author: "Michael Romero A18135877"
format: html
---

## AlphaFold DB

The EBI maintains the largest databse of AlphaFold structure prediction models at: https://alphafold.ebi.ac.uk/ 

From last class (before Halloween) we saw that the PDB had 244,290 (Oct 2025) protein structures. (CHECK FROM CLASS 9)

The total number of protein sequences in UniProtKB is 199,579,901

>**Key Point**: THis is a tiny fraction of sequence space that has structural coverage (0.12%).

```{r}
244290/199579901 *100
```

AFDB is attempting to address this gap...

There are two "Quality Scores" from AlphaFold one for residues (i.e. each amino acid) called **pLDDT** score. The other **PAE** score that measures the confidence in the relative position of two residues (i.e. a score for every pair of residues).

##Generating your own structure predictions

Figure of 5 generated HIV-PR models


And the top ranked model colored by chian
![](HIV)


pLDDT score for model 1 

![](HIVPR_DIMER_23119_0_UNRELAXED_RANK_001_ALPHAFOLD2_MULTIMER_V3_MODEL_4_SEED_000.PDB-HIVPR_DIMER_23119_0_UNRELAXED_RANK_002_ALPHAFOLD2_MULTIMER_V3_MODEL_1_SEED_000.PDB-HIVPR_DIMER_23119_0_UNRELAXED_RANK_003_ALPHAFOLD2_MULTIMER_V3_MODEL_5_SEED.png)

and model 5

![](HIVPR_DIMER_23119_0_UNRELAXED_RANK_001_ALPHAFOLD2_MULTIMER_V3_MODEL_4_SEED_000.PDB-HIVPR_DIMER_23119_0_UNRELAXED_RANK_002_ALPHAFOLD2_MULTIMER_V3_MODEL_1_SEED_000.PDB-HIVPR_DIMER_23119_0_UNRELAXED_RANK_003_ALPHAFOLD2_MULTIMER_V3_MODEL_5_SEED.png)

#Custom analysis of resulting models in R

Read key result files into R. The first thing I need to know is what my results directory/folder is called (i.e. its name is different for every AlphaFold run/job)

```{r}
results.dir <- "HIVPR_dimer_23119_0/"

# File names for all PDB models
pdb_files <- list.files(path=results.dir,
                        pattern="*.pdb",
                        full.names = TRUE)

# Print our PDB file names
basename(pdb_files)

```

```{r}
library(bio3d)

m1 <- read.pdb(pdb_files[1])
m1
```
```{r}
head(m1$atom)
```

```{r}
plot( m1$atom$b[m1$calpha], typ="l" )
```


#Residue consdervation from alignment file

Find the largest AlphaFOld alignment file

```{r}
aln_file <- list.files(path=results.dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file
```
Read this into R

```{r}
aln <- read.fasta(aln_file[1], to.upper = TRUE)

```

How many sequences are in this alignment

```{r}
dim(aln$ali)

```

We can score residue conservation in the alignment with the conserv() function.

```{r}
sim <- conserv(aln)
```

```{r}
pdb <- read.pdb("1hsg")
```

```{r}
plotb3(sim[1:99], sse=trim.pdb(pdb, chain="A"),
       ylab="Conservation Score")
```

```{r}
con<- consensus (aln, cutoff = 0.9)
con$seq
```

